Add agent framework/throttling/hidden model/OS assistant and update conversational search documentation (#6354)
opensearch-project/documentation-website: _search-plugins/conversational-search.md

  - /ml-commons-plugin/conversational-search/
---

# Conversational search

Conversational search allows you to ask questions in natural language and refine the answers by asking follow-up questions. Thus, the conversation becomes a dialog between you and a large language model (LLM). For this to happen, instead of answering each question individually, the model needs to remember the context of the whole conversation. 

Conversational search is implemented with the following components:

- [Conversation history](#conversation-history): Allows an LLM to remember the context of the current conversation and understand follow-up questions.
- [Retrieval-Augmented Generation (RAG)](#rag): Allows an LLM to supplement its static knowledge base with proprietary or current information.

## Conversation history

Conversation history consists of a simple CRUD-like API comprising two resources: _memories_ and _messages_. All messages for the current conversation are kept within one conversation _memory_. A _message_ represents a question/answer pair: a human input question and an artificial intelligence (AI) answer. Messages do not exist by themselves, they must be added to a memory. 

## RAG

> RAG retrieves data from the index and history and sends all the information as context to the LLM. The LLM then supplements its static knowledge base with the dynamically retrieved data. In OpenSearch, RAG is implemented through a search pipeline containing a[retrieval-augmented generation processor]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/rag-processor/). The processor intercepts OpenSearch query results, retrieves previous messages in the conversation from the conversation memory, and sends a prompt to a large language model (LLM). After the processor receives a response from the LLM, it saves the response in conversation memory and returns both the original OpenSearch results of the query and the LLM response. 

Reviewer: ```suggestion
RAG retrieves data from the index and history and sends all the information as context to the LLM. The LLM then supplements its static knowledge base with the dynamically retrieved data. In OpenSearch, RAG is implemented through a search pipeline containing a [retrieval-augmented generation processor]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/rag-processor/). The processor intercepts OpenSearch query results, retrieves previous messages in the conversation from the conversation memory, and sends a prompt to an LLM. After the processor receives a response from the LLM, it saves the response in conversation memory and returns both the original OpenSearch results of the query and the LLM response. 
```

```
{% include copy-curl.html %}

### Step 5: Create a conversation memory

You'll need to create a conversation memory that will store all messages in a conversation. To make the memory easily identifiable, provide a name for the memory in the optional `name` field. This is your only opportunity to name your conversation because the `name` parameter is not updatable:

```json
POST /_plugins/_ml/memory/
{
  "name": "Conversation about NYC population"
}
```
{% include copy-curl.html %}

OpenSearch responds with a memory ID for the newly created memory:

```json
{
  "memory_id": "znCqcI0BfUsSoeNTntd7"
}
```

You'll use the `memory_id` to add messages to the memory. 

> ### Step 6: Use the pipeline for RAG

Reviewer: ```suggestion
### Step 6: Use the RAG pipeline
```

To use the RAG pipeline, send a query to OpenSearch and provide additional parameters in the `ext.generative_qa_parameters` object. 

The `generative_qa_parameters` object supports the following parameters.

Parameter | Required | Description
:--- | :--- | :---
`llm_question` | Yes | The question the LLM must answer. 
`llm_model` | No | Overrides the original model set in the connection in cases where you want to use a different model (for example, GPT 4 instead of GPT 3.5). This option is required if a default model is not set during pipeline creation.

> `memory_id` | No | If you provide a `memory_id`, the pipeline retrieves the 10 most recent messages to the LLM prompt. If you don't specify a `memory_id`, the prior context is not added to the LLM prompt. 

Author: ```suggestion
`memory_id` | No | If you provide a `memory_id`, the pipeline retrieves the 10 most recent messages in the specified memory and adds them to the LLM prompt. If you don't specify a `memory_id`, the prior context is not added to the LLM prompt. 
```

`context_size` | No | The number of search results sent to the LLM. This is typically needed in order to meet the token size limit, which can vary by model. Alternatively, you can use the `size` parameter in the Search API to control the number of search results sent to the LLM.

> `message_size` | No | The number of messages sent to the LLM. Similarly to the number of search results, this affects the total number of tokens seen by the LLM. When not set, the pipeline uses the default message size of `10`.

Author: ```suggestion
`message_size` | No | The number of messages sent to the LLM. Similarly to the number of search results, this affects the total number of tokens received by the LLM. When not set, the pipeline uses the default message size of `10`.
```

        }
      }
    ]
  },
  "ext": {
    "retrieval_augmented_generation": {
      "answer": "The population of the New York City metro area in 2023 is projected to be 18,937,000.",
      "message_id": "x3CecI0BfUsSoeNT9tV9"
    }
  }
}
```
</details>

Now you'll ask an LLM a follow-up question for the same conversation. Again, provide the `memory_id` in the request: 

```json
GET /my_rag_test_data/_search
{
  "query": {
    "match": {
      "text": "What was it in 2022"
    }
  },
  "ext": {
    "generative_qa_parameters": {
      "llm_model": "gpt-3.5-turbo",
      "llm_question": "What was it in 2022",
      "memory_id": "znCqcI0BfUsSoeNTntd7",
      "context_size": 5,
      "message_size": 5,
      "timeout": 15
    }
  }
}
```
{% include copy-curl.html %}

The LLM correctly identifies the conversation subject and returns a relevant response:

```json
{
  ...
  "ext": {
    "retrieval_augmented_generation": {
      "answer": "The population of the New York City metro area in 2022 was 18,867,000.",
      "message_id": "p3CvcI0BfUsSoeNTj9iH"
    }
  }
}
```

> To verify that both messages were added to the memory, provide the `memory_ID` to the Get Message API:

Author: ```suggestion
To verify that both messages were added to the memory, provide the `memory_ID` to the Get Messages API:
```
